0023-bilateral-transfer-protocol/0023-bilateral-transfer-protocol.md

```diff
@@ -175,6 +180,31 @@ ProtocolData ::= SEQUENCE OF SEQUENCE {
 }
 ```
 
+## Authentication
+
+Before anything else, when a client connects to a server, it sends a special
+`Message` request. Its primary `protocolData` entry MUST have name `'auth'`
+and empty data, and among the secondary entries, there MUST be a UTF-8
+`'auth_token'` entry, and a UTF-8 `'auth_username'` entry. The further secondary
+protocol data entries of this `Message` request MAY also be used to send
+additional information to the server. In situations where no authentication
+is needed, the `'auth_token'` and `'auth_username'` data can be set to the
+empty string, but they cannot be omitted.
+
+The server responds with a `Response` or `Error` as appropriate. Again, the
```

Comment:
We should be more specific and list the situations when the server replies with a `Response` or an `Error`. A server replies with `Error` when:
- Any other packet is sent before the auth data.
- The provided authentication is invalid.
- Any of the other protocol rules are violated (e.g. having two subprotos with the same name)

The server sents a response if:
- the provided auth is correct.
- anything else?